common/src/main/java/tachyon/resource/ResourcePool.java

```diff
@@ -66,19 +67,33 @@ protected ResourcePool(int maxCapacity, BlockingQueue<T> resources) {
    */
   public T acquire() {
     // If the resource pool is empty but capacity is not yet full, create a new resource.
-    synchronized (mCapacityLock) {
-      if (mResources.isEmpty() && mCurrentCapacity < mMaxCapacity) {
-        T newResource = createNewResource();
-        mCurrentCapacity ++;
-        return newResource;
-      }
+    T resource = mResources.poll();
+    if (resource != null) {
+      return resource;
+    }
+
+    if (mCurrentCapacity.getAndIncrement() < mMaxCapacity) {
+      return createNewResource();
     }
 
+    mCurrentCapacity.decrementAndGet();
```

Comment:
That's right, At the previous line, the `mCurrentCapacity` is incremented expecting it may be less then the `mMaxCapacity`. But if it's already reached the limit, the increasing need to be rallbacked.